R/colors.R

```diff
@@ -0,0 +1,315 @@
+#'
+#' @param pattern Regex that contains the same number of capturing groups as
+#'   unnamed arguments in `...`. The capturing groups MUST be non-overlapping.
+#' @param ... Functions for decoding each capturing group
+#' @noRd
+match_and_decode <- function(str, pattern, ...) {
+  # Example:
+  # str <- c("#123456", "#ABCDEF")
+  # pattern <- "#([0-9A-F]{2})([0-9A-F]{2})([0-9A-F]{2})"
+  # args <- list(decode_hex, decode_hex, decode_hex)
+
+  args <- rlang::list2(...)
+  m <- regexec(pattern, str)
+  matching_rows <- vapply(m, function(x) { x[1] >= 0 }, logical(1)) # Ex: c(T,T)
+  matches <- regmatches(str[matching_rows], m[matching_rows])
+  if (length(matches) == 0) {
+    return(list(
+      matching_rows = matching_rows
+    ))
+  }
+
+  col_count <- length(matches[[1]])
+  str_matrix <- matrix(unlist(matches), ncol = col_count, byrow = TRUE)
+  # Drop the first column, which is the entire matched string; we only want the
+  # capturing groups
+  str_matrix <- str_matrix[,-1,drop=FALSE]
+  # Ex: str_matrix
+  #      [,1] [,2] [,3]
+  # [1,] "12" "34" "56"
+  # [2,] "AB" "CD" "EF"
+  vals <- lapply(seq_len(ncol(str_matrix)), function(i) {
+    # Ex: decode_hex(c("12", "AB")) => c(18, 171)
+    args[[i]](str_matrix[,i])
+  })
+  results <- do.call(cbind, vals)
+  # Ex: results
+  #      [,1] [,2] [,3]
+  # [1,]   18   52   86
+  # [2,]  171  205  239
+
+  return(list(
+    matching_rows = matching_rows,
+    values = results
+  ))
+}
+
+decode_optional <- function(func, default_value) {
+  force(func)
+  force(default_value)
+
+  function(str) {
+    result <- rep_len(default_value, length(str))
+    has_value <- nzchar(str, keepNA = FALSE) & !is.na(str)
+    result[has_value] <- func(str[has_value])
+    result
+  }
+}
+
+# Takes a vector of strings whose elements represent a SINGLE hex color channel
+# (one or two hexadecimal digits) and return an integer in the range [0-255].
+decode_hex <- function(str) {
+  stopifnot(all(nchar(str) %in% c(1,2)))
+
+  # Single hex digits get doubled up
+  str <- ifelse(nchar(str) == 1, paste0(str, str), str)
+
+  res <- strtoi(str, 16)
+  stopifnot(!any(is.na(res)))
+  res
+}
+
+# Convert strings of floating point numbers [0-255] to integer values in the
+# same range. Valid values outside the range will be clamped. Invalid values
+# will raise errors.
+decode_float_255 <- function(str) {
+  as.integer(pmax(0, pmin(255, round(decode_float_identity(str)))))
+}
+
+# Convert strings of floating point numbers [0-1] to integer values [0-255].
+# Valid values outside the range will be clamped. Invalid values will raise
+# errors.
+decode_float_1 <- function(str) {
+  as.integer(pmax(0, pmin(255, round(decode_float_identity(str) * 255))))
+}
+
+decode_float_identity <- function(str) {
+  stopifnot(all(grepl(paste0("^", regex_float, "$"), str)))
+  as.numeric(str)
+}
+
+encode_hex <- function(values) {
+  if (ncol(values) == 0 || nrow(values) == 0) return(character(0))
```

Comment:
Is this supposed to cover the case where `match_and_decode()` has no matches (and thus returns no `values`)? In that case, `values` will be `NULL`, so perhaps this should be something like `if (!length(values)) return(character(0))`? 

Also, to make it more obvious what sort of input these encoding functions are expecting, perhaps you could also add `if (!is.matrix(values)) stop("Expected a matrix")` after the length check?